### PR TITLE
[Merged by Bors] - feat(Topology/Baire): add basics of Baire measurability

### PR DESCRIPTION
**Pull request by Felix-Weilacher**
Add the notion of Baire measurability for sets in a topological space, analogous to null measurability for sets in a measure space. 

A follow-up PR will add the Kuratowski-Ulam theorem.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/Baire/BaireMeasurable.lean [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+theorem congr (hs : BaireMeasurableSet s) (h : s =ᵇ t) : BaireMeasurableSet t :=
+  EventuallyMeasurableSet.congr (α := α) hs h.symm
+
+end BaireMeasurableSet
+
+open Filter
+
+/--Any Borel set differs from some open set by a meager set. -/
+theorem MeasurableSet.residualEq_open [MeasurableSpace α] [BorelSpace α] (h : MeasurableSet s) :
```
Comment (urkud) [inline Mathlib/Topology/Baire/BaireMeasurable.lean:127]:
```suggestion
theorem MeasurableSet.residualEq_isOpen [MeasurableSpace α] [BorelSpace α] (h : MeasurableSet s) :
```

File: Mathlib/Topology/Baire/BaireMeasurable.lean [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  exact ⟨u, hu, hst.trans htu⟩
+
+section Map
+
+open Set
+
+variable {f : α → β} (hc : Continuous f) (ho : IsOpenMap f)
+
+/-- The preimage of a `Dense` set under a continuous open map is `Dense`. -/
+theorem Dense.dense_preimage_of_isOpenMap {t : Set β} (ht : Dense t) : Dense (f ⁻¹' t) := by
```
Comment (urkud) [inline Mathlib/Topology/Baire/BaireMeasurable.lean:158]: This belongs in `Topology/Maps`.

File: Mathlib/Topology/Baire/BaireMeasurable.lean [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+  rw [dense_iff_closure_eq, ← ho.preimage_closure_eq_closure_preimage hc,
+    dense_iff_closure_eq.mp ht, preimage_univ]
+
+theorem residual_map_le_of_isOpenMap : (residual α).map f ≤ residual β := by
```
Comment (urkud) [inline Mathlib/Topology/Baire/BaireMeasurable.lean:162]: This is called `Filter.Tendsto f (residual α) (residual β)`.

File: Mathlib/Topology/Baire/BaireMeasurable.lean [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+/-- The preimage of a `BaireMeasurableSet` under a continuous open map is Baire measurable. -/
+theorem BaireMeasurableSet.preimage {s : Set β} (h : BaireMeasurableSet s) :
+    BaireMeasurableSet (f⁻¹' s) := by
+  rcases h with ⟨u, hu, hsu⟩
+  refine ⟨f ⁻¹' u, ?_, hsu.filter_mono <| residual_map_le_of_isOpenMap hc ho⟩
+  borelize α β
+  exact hc.measurable hu
+
+theorem Homeomorph.residual_map_eq (h : α ≃ₜ β) : (residual α).map h = residual β := by
```
Comment (urkud) [inline Mathlib/Topology/Baire/BaireMeasurable.lean:179]: This belongs in the file where we prove other basic properties of `residual` (if `Homeomorph` is available there).

File: Mathlib/Topology/Baire/BaireMeasurable.lean [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+  borelize α
+  rcases h with ⟨t, ht, hst⟩
+  rcases ht.residualEq_open with ⟨u, hu, htu⟩
+  exact ⟨u, hu, hst.trans htu⟩
+
+section Map
+
+open Set
+
+variable {f : α → β} (hc : Continuous f) (ho : IsOpenMap f)
```
Comment (urkud) [inline Mathlib/Topology/Baire/BaireMeasurable.lean:155]: As @sgouezel told me a few years ago, using `variable`s for explicit arguments is not a good style. It's not hard to copy+paste them.

File: Mathlib/Topology/Baire/BaireMeasurable.lean (hunk 4 above)
Comment (Felix-Weilacher) [reply Mathlib/Topology/Baire/BaireMeasurable.lean:179]: Currently this is Topology/GDelta, which is strange. What do you think of adding in this PR another new file, say Topology/Baire/Basic, where the definitions and basic properties of `isMeager` go

Comment (Felix-Weilacher) [reply Mathlib/Topology/Baire/BaireMeasurable.lean:176]: (Oh and `Homeomorph is in fact not available there)

File: Mathlib/Topology/Basic.lean [hunk 6]
```diff
@@ -1229,6 +1229,13 @@ theorem Dense.inter_nhds_nonempty (hs : Dense s) (ht : t ∈ 𝓝 x) :
   let ⟨U, hsub, ho, hx⟩ := mem_nhds_iff.1 ht
   (hs.inter_open_nonempty U ho ⟨x, hx⟩).mono fun _y hy => ⟨hy.2, hsub hy.1⟩
 
+/-- The union of a set and the complement of its closure is a dense set. -/
+theorem union_compl_closure_dense {s : Set X} :
+    Dense (s ∪ (closure s)ᶜ) := by
```
Comment (urkud) [inline Mathlib/Topology/Basic.lean:1234]: Is it the same as the recently added `coborder`?

File: Mathlib/Topology/Baire/BaireMeasurable.lean (hunk 4 above)
Comment (urkud) [reply Mathlib/Topology/Baire/BaireMeasurable.lean:176]: Does `Baire/Lemmas` have the required imports?

File: Mathlib/Topology/Maps/Basic.lean [hunk 7]
```diff
@@ -363,6 +363,12 @@ theorem of_isEmpty [h : IsEmpty X] (f : X → Y) : IsOpenMap f := of_nhds_le h.e
 
 end IsOpenMap
 
+/-- The preimage of a `Dense` set under a continuous open map is `Dense`. -/
+theorem Dense.dense_preimage_of_isOpenMap (hf : IsOpenMap f) (hfc : Continuous f)
+    {t : Set Y} (ht : Dense t) : Dense (f ⁻¹' t) := by
```
Comment (urkud) [inline Mathlib/Topology/Maps/Basic.lean:368]: You don't need continuity here, see #15552

File: Mathlib/Topology/Maps/Basic.lean [hunk 8]
```diff
@@ -16,7 +16,7 @@ This file introduces the following properties of a map `f : X → Y` between top
 (Open and closed maps need not be continuous.)
 
 * `Inducing f` means the topology on `X` is the one induced via `f` from the topology on `Y`.
-  These behave like embeddings except they need not be injective. Instead, points of `X` which
+  These behave like embeddings except they need not be injective. Instead, points of `X` whichq
```
Comment (urkud) [inline Mathlib/Topology/Maps/Basic.lean:19]:
```suggestion
  These behave like embeddings except they need not be injective. Instead, points of `X` which
```

File: Mathlib/Topology/Baire/BaireMeasurable.lean [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+  rcases ht.residualEq_isOpen with ⟨u, hu, htu⟩
+  exact ⟨u, hu, hst.trans htu⟩
+
+section Map
+
+open Set
+
+variable {f : α → β}
+
+theorem residual_map_le_of_isOpenMap (hc : Continuous f) (ho : IsOpenMap f) :
```
Comment (urkud) [inline Mathlib/Topology/Baire/BaireMeasurable.lean:157]:
```suggestion
theorem tendsto_residual_of_isOpenMap (hc : Continuous f) (ho : IsOpenMap f) :
```

File: Mathlib/Topology/Basic.lean (hunk 6 above)
Comment (Felix-Weilacher) [reply Mathlib/Topology/Basic.lean:1234]: Looks like it.

File: Mathlib/Topology/Baire/BaireMeasurable.lean (hunk 4 above)
Comment (Felix-Weilacher) [reply Mathlib/Topology/Baire/BaireMeasurable.lean:176]: No. The only import of that file is `Topology/GDelta`

File: Mathlib/Topology/Maps/Basic.lean (hunk 7 above)
Comment (Felix-Weilacher) [reply Mathlib/Topology/Maps/Basic.lean:372]: Thanks, I merged #15552 and deleted my version

File: Mathlib/Topology/Baire/BaireMeasurable.lean [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+import Mathlib.MeasureTheory.Constructions.BorelSpace.Basic
+
+/-!
+# Baire category and Baire measurable sets
+
+This file defines some of the basic notions of Baire category and Baire measurable sets.
+
+## Main definitions
+
+First, we define the notation `=ᵇ`. The denotes eventual equality with respect to the filter of
```
Comment (sgouezel) [inline Mathlib/Topology/Baire/BaireMeasurable.lean:18]:
```suggestion
First, we define the notation `=ᵇ`. This denotes eventual equality with respect to the filter of
```

File: Mathlib/Topology/Basic.lean [hunk 11]
```diff
@@ -1229,6 +1229,19 @@ theorem Dense.inter_nhds_nonempty (hs : Dense s) (ht : t ∈ 𝓝 x) :
...
+  rw [coborder, compl_eq_comm, compl_union, compl_compl, inter_comm]
+  rfl
+
+--Move to Toplogy.LocallyClosed?
+/-- The coborder of any set is dense -/
+theorem dense_coborder {s : Set X} :
+    Dense (coborder s) := by
+  rw [dense_iff_closure_eq, coborder_eq_union_closure_compl, closure_union, ← univ_subset_iff]
+  refine _root_.subset_trans ?_ (union_subset_union_right _ (subset_closure))
+  simp
```
Comment (sgouezel) [inline Mathlib/Topology/Basic.lean:1243]: Could you move them if this is possible, or remove the comment that they should be moved?

File: Mathlib/Topology/Baire/BaireMeasurable.lean [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+
+/-- Notation to say that a property of points in a topological space holds
+almost everywhere in the sense of Baire category. That is, on a residual set. -/
+notation3 "∀ᵇ "(...)", "r:(scoped p => Filter.Eventually p <| residual _) => r
+
+/-- Notation to say that a property of points in a topological space holds on a non meager set. -/
+notation3 "∃ᵇ "(...)", "r:(scoped p => Filter.Frequently p <| residual _) => r
+
+/-- A type tag for `α` with `MeasurableSet` given by `BaireMeasurableSet`. -/
+def BaireMeasurableSpace : Type _ := α
```
[the remark this answers is not in the document]
Comment (Felix-Weilacher) [reply Mathlib/Topology/Baire/BaireMeasurable.lean:51]: Ok, sure. I was just copying what was done for `NullMeasurable`. Do you know why the type synonym is used there?

**Comment by urkud** [comment]
Thanks! :tada:
bors merge

**Comment by Felix-Weilacher** [comment]
I think I fixed this. How do I send it back to bors?

@urkud

**Comment by urkud** [comment]
bors d+
bors merge

**Comment by Felix-Weilacher** [comment]
> ✌️ Felix-Weilacher can now approve this pull request. To approve and merge a pull request, simply reply with `bors r+`. More detailed instructions are available [here](https://bors.tech/documentation/getting-started/#reviewing-pull-requests).

bors r+

**Pull request closed without merging**